### ISSUE TITLE
Switch from mypy to pyright for static type checking

### ISSUE THREAD

**Issue by namurphy** [PlasmaPy/PlasmaPy#2451]
### Description of improvement

We recently added [mypy](https://mypy.readthedocs.io/en/stable/) as a static type checker in our suite of continuous integration tests. This issue proposes to switch to [**pyright**](https://github.com/microsoft/pyright) for static type checking instead.

### Motivation

Before adding mypy to the CI suite, I didn't really look into alternatives to mypy.  From the information discussions I've seen online, generally there seems to be a preference for pyright over mypy with respect to quality of error messages, performance, quality of the code base, and responsiveness of the maintainers to bug reports.  The quality of error messages is particularly important because we have a lot of new contributors.

### Implementation strategy

The documentation for pyright discusses [getting started](https://microsoft.github.io/pyright/#/getting-started).

 - [ ] Merge or close all PRs in progress that are adding type hint annotations.
 - [ ] Create a tox environment for pyright (see also #2431)
 - [ ] Create a pyright configuration that ignores all existing errors (see also #2424)
 - [ ] Add the tox environment to GitHub Actions (see also #2432)
 - [ ] Update or delete `# type: ignore[mypy-error-code]` comments
 - [ ] In the setting for the GitHub repo, make the pyright check required for new PRs.

### Additional context

Original issue for mypy: #268.

**Comment by namurphy** [comment]
I'm labeling this as "on hold" for the moment, since it's something we'd need to consider more thoughtfully.

**Comment by namurphy** [comment]
Some initial impressions after trying out pyright:

 - I had difficulty installing pyright. It ended up working when I did `conda install pyright`.  
 - I find the mypy documentation to be easier to glean information from.  
 - mypy has a customization option that will print out the code and show the location of the error, which I found quite helpful.  I didn't see an option for pyright.
 - pyright doesn't have a plugin system, though I don't think that would affect us too much.
 - pyright seems pretty thorough. It found 1306 errors in "basic" mode, excluding `tests` directorires.  
 - There should probably be only one static type checker in our CI since the `type: ignore` comments have different error codes.  However, if we're going through the process of adding type hint annotations in a file, it might be worth both running pyright and mypy. So, maybe it's worth adding a tox environment for pyright too?
 - Enabling a static type checker on a large existing code base is going to be a pain regardless of which one you choose. 🙃

I'm not convinced yet that it would be worth going through the effort to switch.
